modules/block/src/main/kotlin/jp/co/soramitsu/iroha2/Genesis.kt

```diff
@@ -36,12 +51,110 @@ open class Genesis(open val genesisBlock: RawGenesisBlock) {
+
+            return Genesis(RawGenesisBlock(listOf(GenesisTransaction(uniqueIsi.mergeMetadata()))))
         }
 
-        /**
-         * Return empty genesis
-         */
-        fun getEmpty() = Genesis(RawGenesisBlock(listOf(GenesisTransaction(listOf()))))
+        private fun MutableSet<Instruction>.mergeMetadata(): List<Instruction> {
+            val metadataMap = mutableMapOf<Any, Metadata>()
+
+            // only for Instruction.Register
+            this.extractIdentifiableBoxes().forEach { idBox ->
+                metadataMap.putMergedMetadata(idBox)
+            }
+            this.findIsiToReplace(metadataMap).forEach { (metadata, toReplace) ->
+                toReplace.forEach { isi -> this.remove(isi) }
+
+                val idBox = toReplace.first().extractIdentifiableBox()
+                val registrableBox = idBox?.toRegisterBox(metadata)
+                    ?: throw RuntimeException("IdentifiableBox shouldn't be null")
+                this.add(Instruction.Register(RegisterBox(registrableBox.evaluatesTo())))
+            }
+
+            return this.sorted()
+        }
+
+        private fun MutableSet<Instruction>.sorted() = this.sortedWith(
+            compareByDescending { instruction ->
+                when (instruction) {
+                    is Instruction.Register -> when (instruction.extractIdentifiableBox()) {
+                        is IdentifiableBox.NewDomain -> 6
+                        is IdentifiableBox.NewAccount -> 5
+                        is IdentifiableBox.NewAssetDefinition -> 4
+                        is IdentifiableBox.PermissionTokenDefinition -> 3
+                        is IdentifiableBox.NewRole -> 2
+                        else -> 1
+                    }
+
+                    else -> -1
+                }
+            }
+        )
+
+        private fun IdentifiableBox.toRegisterBox(metadata: Metadata) = when (this) {
+            is IdentifiableBox.NewAccount -> RegistrableBox.Account(
+                NewAccount(this.newAccount.id, this.newAccount.signatories, metadata)
+            )
+
+            is IdentifiableBox.NewDomain -> RegistrableBox.Domain(
+                NewDomain(this.newDomain.id, this.newDomain.logo, metadata)
+            )
+
+            is IdentifiableBox.NewAssetDefinition -> RegistrableBox.AssetDefinition(
+                NewAssetDefinition(
+                    this.newAssetDefinition.id,
+                    this.newAssetDefinition.valueType,
+                    this.newAssetDefinition.mintable,
+                    metadata
+                )
+            )
+
+            else -> throw IrohaSdkException("Unexpected type ${this::class}")
+        }
+
+        private fun MutableMap<Any, Metadata>.putMergedMetadata(idBox: IdentifiableBox) {
+            fun MutableMap<Any, Metadata>.putOrMerge(id: Any, metadata: Metadata) = when (this[id]) {
+                null -> this[id] = metadata
+                else -> this[id] = this[id]!!.merge(metadata)
```

Comment:
what if there is the same key but different values? shouldn't we go with an exception?